clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -1034,6 +1121,21 @@ private TransactionalRequestResult handleCachedTransactionRequestResult(
         return pendingResult;
     }
 
+    private boolean canBumpEpoch() {
+        return !isTransactional() ||
+                apiVersions.get(transactionCoordinator.idString()).apiVersion(ApiKeys.INIT_PRODUCER_ID).maxVersion >= 3;
```

Comment:
We seem to be missing a couple null checks here.

Reply:
Fixed the null checks. But there's also an issue where we can lose connection to the coordinator, and wind up calling this method before we've found the new coordinator. Need to think a bit about how to handle that.